### PR TITLE
MPD noidle and command lists (refactored version of PR #1612)

### PR DESCRIPTION
**Pull request by ejurgensen**

### CODE REVIEW

**Comment by ejurgensen** [comment]
@grobian if you are up to testing these changes it would be great. I don't use the mpd part of OwnTone regularly so I can't properly test.

File: src/mpd.c [hunk 1]
```diff
@@ -30,10 +30,6 @@
 #include <limits.h>
 #include <errno.h>
 #include <pthread.h>
-#include <sys/param.h>
-#include <sys/queue.h>
-#include <sys/types.h>
-#include <stdint.h>
```
Comment (grobian) [inline src/mpd.c:36]: doesn't this belong in its own commit as "cleanup"?

File: src/mpd.c [hunk 2]
```diff
@@ -85,7 +80,40 @@ static bool mpd_plugin_httpd;
...
+/**
+ * config:
+ * max_command_list_size KBYTES
+ * The maximum size a command list. Default is 2048 (2 MiB).
+ *
+ * https://github.com/MusicPlayerDaemon/MPD/blob/master/src/client/Config.cxx
+ */
+#define CLIENT_MAX_COMMAND_LIST_DEFAULT		(2048*1024)
+
+static struct {
```
Comment (grobian) [inline src/mpd.c]: this is an anonymous struct for a single member, is it expected to contain more in the future?  How about the current static vars such as allow_modifying_stored_playlists, shouldn't those be in this structure too (as a separate commit)?

Also, probably better to declare this structure named, and then define mpd_config as static variable.

(Config is very generic easy to clash, and IMO inconsistent with the rest of the file

File: src/mpd.c [hunk 3]
```diff
@@ -368,31 +434,50 @@ mpd_time(char *buffer, size_t bufferlen, time_t t)
...
+ * For "window START:END" The end index can be omitted, which means the range is open-ended.
  */
 static int
-mpd_pars_range_arg(char *range, int *start_pos, int *end_pos)
+mpd_pars_range_arg(const char *range, int *start_pos, int *end_pos)
 {
   int ret;
 
-  if (strchr(range, ':'))
+  static char separator = ':';
```
Comment (grobian) [inline src/mpd.c]: yeah... why not `#define MPD_SEPARATOR ':'` instead

File: src/mpd.c [hunk 4]
```diff
@@ -368,31 +434,50 @@ mpd_time(char *buffer, size_t bufferlen, time_t t)
...
+
+  if (sep_pos)
     {
-      ret = sscanf(range, "%d:%d", start_pos, end_pos);
-      if (ret < 0)
-	{
-	  DPRINTF(E_LOG, L_MPD, "Error parsing range argument '%s' (return code = %d)\n", range, ret);
-	  return -1;
-	}
+      *sep_pos++ = '\0';
```
Comment (grobian) [inline src/mpd.c]: I've learned in another PR that this is preferred to be two operations instead of one

**Review by grobian — COMMENTED** [review]
meh, sorry, I don't know what I'm looking really

Time for a testsuite perhaps?  Easy to interact with nc and track outputs as I've been doing to test some responses in a unit-test fashion.

**Comment by grobian** [comment]
@ejurgensen to say something about this it needs to be more pinned down to commits that do something, at least for me.  I guess now it's just running clients against it and YOLO.

I've been playing with the thought to add some testsuite script, because while testing I noticed various things not working properly or not at all, and with something like https://mpd.readthedocs.io/en/latest/protocol.html#id34 in hand, it should be able to craft small sequence-based tests and verified outputs.  Client compatibility I don't know how to address in a test suite, but having at least something here would be a start.

**Comment by ejurgensen** [comment]
Thanks for actually looking at the code, I didn't expect that. However, it seems you looked at the commits one at a time, so some of your comments are outdated (as Github also shows).

> I guess now it's just running clients against it and YOLO.

Indeed it is :-) I think a test suite would be fine, but then an actual C test framework needs to be chosen and added, and tbh I never bothered to do that (and in my defense, nor did the previous maintainers).

**Comment by grobian** [comment]
No need to overdo it, just verifying the protocol is a good start.

Owntone doesn't even work with mpc, that's why I'm trying to implement the features leading up to a recent protocol version.

This commit conflicts with that effort via my PRs, so I think we're looking at what you're going to do next eagerly to see who needs to move in what direction.

**Comment by ejurgensen** [comment]
> Owntone doesn't even work with mpc, that's why I'm trying to implement the features leading up to a recent protocol version

Seems to work ok on my computer with mpc 0.35. What problem are you seeing with mpc? 

> This commit conflicts with that effort via my PRs, so I think we're looking at what you're going to do next eagerly to see who needs to move in what direction.

I'll try to merge your 0.23 changes first, and then fix conflicts with this branch

**Comment by grobian** [comment]
on (a slightly tinkered of) the released version of owntone:
```
% mpc
warning: MPD 0.21 required
volume: 50%   repeat: off   random: off   single: off   consume: off
%
```

**Comment by ejurgensen** [comment]
Ok, it wasn't clear that you were referring to the release version

File: src/mpd.c
Comment (grobian) [inline src/mpd.c:1]: this commit seems rather clean and focussed, can we have it cherry-picked to master/main?
